File: backend/python/app/agents/agent_loop/protocol/agui.py
"""AG-UI wire vocabulary and framing helper.

Event names match AG-UI's CURRENT spec (https://docs.ag-ui.com/concepts/events)
— in particular the `REASONING_*` suite, not the `THINKING_*` names AG-UI
removed in its 1.0 release.
"""

from __future__ import annotations

import uuid
from enum import Enum
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from fastapi import Request


class AGUIEventType(str, Enum):
    RUN_STARTED = "RUN_STARTED"
    RUN_FINISHED = "RUN_FINISHED"
    RUN_ERROR = "RUN_ERROR"
    STEP_STARTED = "STEP_STARTED"
    STEP_FINISHED = "STEP_FINISHED"

    TEXT_MESSAGE_START = "TEXT_MESSAGE_START"
    TEXT_MESSAGE_CONTENT = "TEXT_MESSAGE_CONTENT"
    TEXT_MESSAGE_END = "TEXT_MESSAGE_END"

    REASONING_START = "REASONING_START"
    REASONING_MESSAGE_START = "REASONING_MESSAGE_START"
    REASONING_MESSAGE_CONTENT = "REASONING_MESSAGE_CONTENT"
    REASONING_MESSAGE_END = "REASONING_MESSAGE_END"
    REASONING_END = "REASONING_END"

    TOOL_CALL_START = "TOOL_CALL_START"
    TOOL_CALL_ARGS = "TOOL_CALL_ARGS"
    TOOL_CALL_END = "TOOL_CALL_END"
    TOOL_CALL_RESULT = "TOOL_CALL_RESULT"

    STATE_DELTA = "STATE_DELTA"
    STATE_SNAPSHOT = "STATE_SNAPSHOT"

    CUSTOM = "CUSTOM"
    HEARTBEAT = "HEARTBEAT"


def new_id(prefix: str) -> str:
    """Short, prefixed id for AG-UI `messageId`/`toolCallId` fields where
    no upstream identity already exists (`ToolCall.id` is reused directly
    when present — see `agui_emitter.py`)."""
    return f"{prefix}_{uuid.uuid4().hex[:12]}"


def frame(event_type: AGUIEventType, **fields: Any) -> dict[str, Any]:
    """One AG-UI event, hybrid-framed as PipesHub's existing `{event, data}`
    SSE dict: `event` carries the AG-UI type name (so Node's line-based
    `event:` interception and the frontend's `parseSSEBuffer`, both of
    which require an `event:` line, keep working unchanged) while `data`
    carries the FULL AG-UI-compliant JSON, including `"type"` — the
    payload itself is byte-for-byte what a stock `@ag-ui/client` expects
    once the redundant `event:` line is dropped. A deliberate deviation
    from the stock `ag-ui-protocol` `EventEncoder` (data-only frames) —
    see the migration plan's "Wire framing decision" section. Returning
    this exact `{"event", "data"}` shape means `stream_bridge.py`'s
    queue/yield loop needs no protocol-specific branch at all.
    """
    return {"event": event_type.value, "data": {"type": event_type.value, **fields}}


def resolve_protocol(protocol_field: str | None, request: "Request") -> str:
    """Resolve the SSE wire protocol shared by every `/chat/stream`-shaped
    route (`agent.py::chat_stream`, `chatbot.py::askAIStream`). AG-UI is the
    only supported protocol: this always returns `"agui"`, independent of
    the request body field or `?protocol=` query param — there is no
    legacy fallback to negotiate into. Signature kept unchanged (still
    takes `protocol_field`/`request`) so call sites don't need to change if
    protocol negotiation is ever reintroduced."""
    del protocol_field, request
    return "agui"


__all__ = ["AGUIEventType", "new_id", "frame", "resolve_protocol"]

File: backend/nodejs/apps/src/modules/enterprise_search/utils/agui.ts
/**
 * AG-UI protocol helpers for the Node.js proxy layer — see the AG-UI
 * migration plan. Node intercepts a handful of event names to persist
 * conversations to Mongo, keying off the AG-UI names
 * (`RUN_FINISHED`/`RUN_ERROR`/`CUSTOM{name:"ask_user_question"}`), and
 * Node's own outbound frames (the early `connected` event, the
 * re-emitted `complete`) are reframed to `CUSTOM`/`RUN_FINISHED` too.
 *
 * AG-UI is the only supported wire protocol; the legacy SSE format has
 * been removed. `LEGACY_PROTOCOL` and the (now unreachable) legacy
 * formatting branches remain in this proxy layer purely as dead code to
 * minimize churn — no caller can select them anymore. See
 * `resolveProtocol()`.
 */

export const AGUI_PROTOCOL = 'agui' as const;
export const LEGACY_PROTOCOL = 'legacy' as const;
export type SSEProtocol = typeof AGUI_PROTOCOL | typeof LEGACY_PROTOCOL;

/**
 * AG-UI is the only supported protocol — always resolve to it regardless
 * of what the caller sends. Kept as a function (rather than inlining the
 * constant at call sites) so the protocol can be reintroduced as a real
 * negotiation later without touching every call site.
 */
export const resolveProtocol = (
  _body?: Record<string, unknown>,
  _query?: Record<string, unknown>,
): SSEProtocol => AGUI_PROTOCOL;

export const isAGUI = (protocol: SSEProtocol | undefined): boolean =>
  protocol === AGUI_PROTOCOL;

/**
 * One AG-UI event, hybrid-framed exactly like the Python side's
 * `protocol/agui.py::frame()` — `event:` line carries the AG-UI type
 * name (so this file's own line-based parsing, and the frontend's
 * `parseSSEBuffer`, keep working unchanged), `data:` carries the full
 * AG-UI JSON including `"type"`.
 */
export const frameAGUI = (
  type: string,
  fields: Record<string, unknown> = {},
): string => `event: ${type}\ndata: ${JSON.stringify({ type, ...fields })}\n\n`;

/** AG-UI type names this proxy layer needs to recognize on the wire. */
export const AGUIEventType = {
  RUN_STARTED: 'RUN_STARTED',
  RUN_FINISHED: 'RUN_FINISHED',
  RUN_ERROR: 'RUN_ERROR',
  STATE_SNAPSHOT: 'STATE_SNAPSHOT',
  CUSTOM: 'CUSTOM',
} as const;
